rasp/tor_engine.py

```diff
@@ -1,57 +1,109 @@
+            can also be set with the ``RASP_TOR_CONTROL_PASSWORD``
+            environment variable.
+        control_password (str, optional): Password to protect control
+            port usage,
+            can also be set with the ``RASP_TOR_CONTROL_PORT``
+            environment variable.
+     """
+    def __init__(self,
+                 headers=None,
+                 address=None,
+                 port=None,
+                 control_port=None,
+                 control_password=None):
+        super(TorEngine, self).__init__(headers)
 
-        proxy_opener = urllib.request.build_opener(self.proxy_handler)
-        urllib.request.install_opener(proxy_opener)
+        self.address = (
+            address
+            or os.environ.get('RASP_TOR_ADDRESS')
+            or '127.0.0.1'
+        )
+        self.port = (
+            port
+            or os.environ.get('RASP_TOR_PORT')
+            or 9050
+        )
+        self.control_password = (
+            control_password
+            or os.environ.get('RASP_TOR_CONTROL_PASSWORD')
+            or getpass.getpass("Tor control password: ")
+        )
+        self.control_port = (
+            control_port
+            or os.environ.get('RASP_TOR_CONTROL_PORT')
+            or 9051
+        )
+
+        proxy_uri = 'socks5://{}:{}'.format(self.address, self.port)
+        proxies = {'http': proxy_uri, 'https': proxy_uri}
+        self.session.proxies.update(proxies)
 
     def __copy__(self):
         return TorEngine(
-            self.pw,
-            self.control,
-            self.signal,
-            self.proxy_handler,
             self.data,
-            self.headers
+            self.headers,
+            self.address,
+            self.port,
+            self.control_port,
+            self.control_password,
         )
 
-    def send_signal(self):
-        conn = stem.connection.connect(
-            control_port=self.control,
-            password=self.pw
-        )
-        conn.signal(self.signal)
-        conn.close()
-
-    def get_page_source(self, url):
-        try:
-            self.send_signal()
-            req = urllib.request.Request(url, self.data, self.headers)
-            res = urllib.request.urlopen(req)
-
-            try:
-                return Webpage(url, str(res.read()))
-            except http.client.IncompleteRead:
-                return
-        except urllib.error.HTTPError as e:
-            time.sleep(2)
+    def clean_circuits(self):
+        """Refreshes the mapping of nodes we connect through.
+
+        When a new path from the user to the destination is required, this
+        method refreshes the node path we use to connect. This gives us a new IP
+        address with which the destination sees."""
+        kwargs = {'address': self.address, 'port': self.control_port}
+        with Controller.from_port(**kwargs) as controller:
+            controller.authenticate(password=self.control_password)
```

Comment:
I do like the context manager syntax a lot, but would caution not to make that the only way to use the engine.  A pretty straight forward pattern is to leave something like `cleanup` and call that in the `exit`, so that you can do either context manager style or the "normal" way.  Ideally in both cases the user does nothing to clean up, so I'd say `__del__` and `__exit__` both calling a `__cleanup()` or something like that would make most sense.

Also for the `__del__` - `__exit__` clash, my understanding is that `__del__` would only get called when the refcount dropped to zero regardless, so I don't think they practically can clash, unless you explicitly called `del foo` on the object after the context manager had closed.  I might be off on that though.